### ISSUE TITLE
BUG: can not add file

### ISSUE THREAD

**Issue by fundon** [nvim-neo-tree/neo-tree.nvim#1179]
### Did you check docs and existing issues?

- [X] I have read all the docs.
- [X] I have searched the existing issues.
- [X] I have searched the existing discussions.

### Neovim Version (nvim -v)

NVIM v0.9.2

### Operating System / Version

NixOS 23.05.4076.8a4c17493e5c (Stoat)

### Describe the Bug

Error when adding file and file has a problem with permission.

### Screenshots, Traceback

<img width="937" alt="Screenshot 2023-10-12 at 16 59 43" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/166789ec-7295-441d-bd01-06b0315b7aa4">

<img width="587" alt="Screenshot 2023-10-12 at 17 00 05" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/f1fddddd-344d-4225-ab50-5a45d4b6e07a">

### Steps to Reproduce

1. init a NixOS with https://orbstack.dev/ 
2. install neovim `nix-env -iA nixpkgs.neovim`
3. OrbStack will automatically mount  macOS `/Users` to nixOS `/`
4. `ls -la /`

<img width="474" alt="Screenshot 2023-10-12 at 17 11 04" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/e2565064-d109-45aa-aa91-60b10a1a1fe3">

6. `ln -sf /Users/fundon/Documents $HOME`
7.  `cd $HOME/Documents`
8. `touch a.txt`
9.  `nvim a.txt`
10.  add file with `neotree` 
<img width="433" alt="Screenshot 2023-10-12 at 17 12 46" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/f9033766-7183-49df-8e50-cd42e909b252">

### Expected Behavior

`touch file` works

### Your Configuration

```Lua
local lazypath = vim.fn.stdpath "data" .. "/lazy/lazy.nvim"
local uv = vim.uv or vim.loop
if not uv.fs_stat(lazypath) then
  vim.fn.system {
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable", -- latest stable release
    lazypath,
  }
end
vim.opt.runtimepath:prepend(lazypath)

vim.opt.clipboard = "unnamedplus"

-- In OrbStack
if vim.fn.executable('pbcopy') then
  vim.g.clipboard = {
    name = 'OrbStack clipboard',
    copy = {
      ['+'] =  'pbcopy',
      ['*'] =  'pbcopy',
    },
    paste = {
      ['+'] = 'pbpaste',
      ['*'] = 'pbpaste',
    },
    cache_enabled = false,
  }
end

vim.opt.cursorline = true

vim.g.mapleader = " " -- make sure to set `mapleader` before lazy so your mappings are correct
vim.g.maplocalleader = " "
vim.wo.number = true
vim.o.mouse = "a"
vim.o.hlsearch = false
vim.o.ignorecase = true
vim.o.smartcase = true
vim.o.updatetime = 250
vim.wo.signcolumn = "yes"
vim.o.completeopt = "menuone,noselect"
vim.o.breakindent = true
vim.o.undofile = true
vim.o.termguicolors = true

vim.opt.tabstop = 2
vim.opt.shiftwidth = 2
vim.opt.expandtab = true
vim.bo.softtabstop = 2

vim.keymap.set({ "n", "v" }, "<Space>", "<Nop>", { silent = true })
vim.keymap.set("n", "k", "v:count == 0 ? 'gk' : 'k'", { expr = true, silent = true })
vim.keymap.set("n", "j", "v:count == 0 ? 'gj' : 'j'", { expr = true, silent = true })

require("lazy").setup {
  {'VonHeikemen/lsp-zero.nvim', branch = 'v3.x'},

  -- LSP Support
  'williamboman/mason.nvim',
  'williamboman/mason-lspconfig.nvim',
  'neovim/nvim-lspconfig',

  -- Autocompletion
  'hrsh7th/nvim-cmp',
  'hrsh7th/cmp-nvim-lsp',
  'L3MON4D3/LuaSnip',

  'hrsh7th/cmp-buffer',
  'hrsh7th/cmp-cmdline',
  'hrsh7th/cmp-path',

  'saadparwaiz1/cmp_luasnip',

  "folke/which-key.nvim",

  { "folke/neoconf.nvim", cmd = "Neoconf" },

  "folke/neodev.nvim",

  "wbthomason/packer.nvim",

  "savq/melange",

  { 
    "lewis6991/gitsigns.nvim",
    config = function()
      require("gitsigns").setup()
    end,
  },

  "sindrets/diffview.nvim",

  {
    "j-hui/fidget.nvim",
    tag = "legacy",
    config = function()
      require("fidget").setup()
    end,
  },

  {
    "numToStr/Comment.nvim",
    config = function()
      require("Comment").setup()
    end,
  },

  -- tmux
  {
    "aserowy/tmux.nvim",
    keys = {
      {
        "<C-h>",
        "<cmd>lua require('tmux').move_left()<CR>",
        desc = "   window left",
      },
      {
        "<C-l>",
        "<cmd>lua require('tmux').move_right()<CR>",
        desc = "   window right",
      },
      {
        "<C-j>",
        "<cmd>lua require('tmux').move_bottom()<CR>",
        desc = "   window down",
      },
      {
        "<C-k>",
        "<cmd>lua require('tmux').move_top()<CR>",
        desc = "   window up",
      },
    },
    config = function()
      require("tmux").setup {
        copy_sync = {
          -- enables copy sync and overwrites all register actions to
          -- sync registers *, +, unnamed, and 0 till 9 from tmux in advance
          enable = false,

          -- ignore specific tmux buffers e.g. buffer0 = true to ignore the
          -- first buffer or named_buffer_name = true to ignore a named tmux
          -- buffer with name named_buffer_name :)
          ignore_buffers = { empty = false },

          -- TMUX >= 3.2: yanks (and deletes) will get redirected to system
          -- clipboard by tmux
          redirect_to_clipboard = false,

          -- offset controls where register sync starts
          -- e.g. offset 2 lets registers 0 and 1 untouched
          register_offset = 0,

          -- sync clipboard overwrites vim.g.clipboard to handle * and +
          -- registers. If you sync your system clipboard without tmux, disable
          -- this option!
          sync_clipboard = true,

          -- syncs deletes with tmux clipboard as well, it is adviced to
          -- do so. Nvim does not allow syncing registers 0 and 1 without
          -- overwriting the unnamed register. Thus, ddp would not be possible.
          sync_deletes = true,

          -- syncs the unnamed register with the first buffer entry from tmux.
          sync_unnamed = true,
        },
        navigation = {
          -- cycles to opposite pane while navigating into the border
          cycle_navigation = true,

          -- enables default keybindings (C-hjkl) for normal mode
          enable_default_keybindings = false,

          -- prevents unzoom tmux when navigating beyond vim border
          persist_zoom = false,
        },
        resize = {
          -- enables default keybindings (A-hjkl) for normal mode
          enable_default_keybindings = false,

          -- sets resize steps for x axis
          resize_step_x = 1,

          -- sets resize steps for y axis
          resize_step_y = 1,
        },
      }
    end,
  },

  {
    'Lilja/zellij.nvim',
    config = function()
      require('zellij').setup {
        vimTmuxNavigatorKeybinds = true
      }
    end
  },

  {
    "kdheepak/tabline.nvim",
    keys = {
      {
        "<TAB>",
        "<cmd>TablineBufferNext<CR>",
        desc = "goto next buffer",
      },
      {
        "<S-TAB>",
        "<cmd>TablineBufferPrevious<CR>",
        desc = "goto prev buffer",
      },
      {
        "<leader>x",
        "<cmd>bdelete<CR>",
        desc = "close buffer",
      },
    },
    config = function()
      require("tabline").setup {
        -- Defaults configuration options
        enable = true,
        options = {
          -- If lualine is installed tabline will use separators configured in lualine by default.
          -- These options can be used to override those settings.
          section_separators = { "", "" },
          component_separators = { "", "" },
          max_bufferline_percent = 66, -- set to nil by default, and it uses vim.o.columns * 2/3
          show_tabs_always = false, -- this shows tabs only when there are more than one tab or if the first tab is named
          show_devicons = true, -- this shows devicons in buffer section
          show_bufnr = false, -- this appends [bufnr] to buffer section,
          show_filename_only = false, -- shows base filename only instead of relative path in filename
          modified_icon = "+ ", -- change the default modified icon
          modified_italic = false, -- set to true by default; this determines whether the filename turns italic if modified
          show_tabs_only = false, -- this shows only tabs instead of tabs + buffers
        },
      }
      -- Use showtabline in gui vim
      vim.cmd "set guioptions-=e"
      -- store tabpages and globals in session
      vim.cmd "set sessionoptions+=tabpages,globals"
    end,
    dependencies = {
      { "nvim-lualine/lualine.nvim", opt = true },
      { "nvim-tree/nvim-web-devicons", opt = true },
    },
  },

  {
    "nvim-treesitter/nvim-treesitter",
    build = ":TSUpdate",
    event = "BufReadPost",
    ---@type TSConfig
    opts = {
      sync_install = false,
      highlight = { enable = true },
      indent = { enable = true },
      context_commentstring = { enable = true, enable_autocmd = false },
      ensure_installed = {
        "bash",
        "css",
        "html",
        "javascript",
        "json",
        "lua",
        "markdown",
        "markdown_inline",
        "python",
        "query",
        "regex",
        "tsx",
        "typescript",
        "vim",
        "yaml",
        "rust",
        "zig",
      },
    },
    ---@param opts TSConfig
    config = function(plugin, opts)
      -- if plugin.ensure_installed then
      --   require("lazyvim.util").deprecate("treesitter.ensure_installed", "treesitter.opts.ensure_installed")
      -- end
      require("nvim-treesitter.configs").setup(opts)
    end,
  },

  {
    "nvim-lualine/lualine.nvim",
    config = function()
      require("lualine").setup {
        sections = {
          lualine_x = {
            {
              require("lazy.status").updates,
              cond = require("lazy.status").has_updates,
              color = { fg = "#ff9e64" },
            },
          },
        },
      }
    end,
  },

  {
    "nvim-telescope/telescope.nvim",
    tag = "0.1.0",
    dependencies = { { "nvim-lua/plenary.nvim" } },
  },

  {
    "iamcco/markdown-preview.nvim",
    ft = "markdown",
    build = function()
      vim.fn["mkdp#util#install"]()
    end,
  },

  "MunifTanjim/nui.nvim",

  { "nvim-telescope/telescope.nvim", branch = "0.1.x", dependencies = { "nvim-lua/plenary.nvim" } },
  { "nvim-telescope/telescope-fzf-native.nvim", build = "make" },

  {
    "nvim-neo-tree/neo-tree.nvim",
    keys = {
      { "<leader>ft", "<cmd>Neotree toggle<cr>", desc = "NeoTree" },
    },
    branch = "v3.x",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "nvim-tree/nvim-web-devicons", -- not strictly required, but recommended
      "MunifTanjim/nui.nvim",
    },
  },

  "lukas-reineke/indent-blankline.nvim",

  "echasnovski/mini.nvim",

  {
    "roobert/tailwindcss-colorizer-cmp.nvim",
    -- optionally, override the default options:
    config = function()
      require("tailwindcss-colorizer-cmp").setup {
        color_square_width = 2,
      }
    end,
  },

  {
    "goolord/alpha-nvim",
    config = function()
      require("alpha").setup(require("alpha.themes.startify").config)
    end,
  },

  {
    "kylechui/nvim-surround",
    -- tag = "*", -- Use for stability; omit to use `main` branch for the latest features
    config = function()
      require("nvim-surround").setup({
        -- Configuration here, or leave empty to use defaults
      })
    end
  },

  {
    "windwp/nvim-autopairs",
    config = function() require("nvim-autopairs").setup {} end
  },

  {
    'saecki/crates.nvim',
    config = function() require("crates").setup {} end,
    dependencies = {
      "nvim-lua/plenary.nvim",
    }
  },

  {
    'akinsho/toggleterm.nvim',
    version = "*",
    config = true
  },

  -- Themes
  {
    "catppuccin/nvim",
    name = "catppuccin",
    -- lazy = true,
    config = function()
      require("catppuccin").setup {
        flavour = "macchiato", -- mocha, macchiato, frappe, latte
      }
      -- vim.api.nvim_command "colorscheme catppuccin"
    end,
  },
  {
    "projekt0n/github-nvim-theme",
  },
  {
    "folke/tokyonight.nvim",
    lazy = true,
    opts = { style = "moon" },
  },
  {
    "rebelot/kanagawa.nvim",
    lazy = true,
    opts = { theme = "lotus" },
  },
  {
    "mhanberg/output-panel.nvim",
    event = "VeryLazy",
    config = function()
      require("output_panel").setup()
    end
  },
  {
    "karb94/neoscroll.nvim",
    config = function()
      require("neoscroll").setup()
    end
  },
  {
    'stevearc/oil.nvim',
    opts = {},
    -- Optional dependencies
    dependencies = { "nvim-tree/nvim-web-devicons" },
  },
  {
    "folke/edgy.nvim",
    event = "VeryLazy",
    opts = {}
  },
  {
    "folke/flash.nvim",
    event = "VeryLazy",
    ---@type Flash.Config
    opts = {},
    keys = {
      {
        "s",
        mode = { "n", "x", "o" },
        function()
          -- default options: exact mode, multi window, all directions, with a backdrop
          require("flash").jump()
        end,
        desc = "Flash",
      },
      {
        "S",
        mode = { "n", "o", "x" },
        function()
          require("flash").treesitter()
        end,
        desc = "Flash Treesitter",
      },
      {
        "r",
        mode = "o",
        function()
          require("flash").remote()
        end,
        desc = "Remote Flash",
      },
    },
  }
}

local lsp_zero = require('lsp-zero')
lsp_zero.on_attach(function(client, bufnr)
  -- see :help lsp-zero-keybindings
  -- to learn the available actions
  lsp_zero.default_keymaps({buffer = bufnr})
end)

-- see :help lsp-zero-guide:integrate-with-mason-nvim
-- to learn how to use mason.nvim with lsp-zero
require('mason').setup({})
require('mason-lspconfig').setup({
  ensure_installed = {},
  handlers = {
    lsp_zero.default_setup,
  }
})

local lspconfig = require('lspconfig')
lspconfig.rust_analyzer.setup{
  cmd = { 'rust-analyzer' };
}
lspconfig.nil_ls.setup{
  cmd = { 'nil' };
}
lspconfig.lua_ls.setup{
  cmd = { 'lua-language-server' };
}

local cmp = require('cmp')
local cmp_action = lsp_zero.cmp_action()
local cmp_format = lsp_zero.cmp_format()

cmp.setup({
  formatting = cmp_format,
  mapping = cmp.mapping.preset.insert({
    -- Scroll up and down in the completion documentation
    ['<C-u>'] = cmp.mapping.scroll_docs(-4),
    ['<C-d>'] = cmp.mapping.scroll_docs(4),
  })
})

vim.cmd("colorscheme kanagawa")

-- The line beneath this is called `modeline`. See `:help modeline`
-- vim: ts=2 sts=2 sw=2 et
```

**Comment by pysan3** [comment]
The following code is basically what neo-tree is doing when creating a file.
Could you check if this lua code works on your machine?

> Open neovim
> Create a file named `tmp.lua`
> Copy the following code
> Run `:so`

If it does not, then it might be a permission issue in your folder.

```lua
local loop = vim.loop
local destination = vim.fn.fnamemodify("~/Documents/b.txt", ":p")
local open_mode = loop.constants.O_CREAT
  + loop.constants.O_WRONLY
  + loop.constants.O_TRUNC
local fd = loop.fs_open(destination, "w", open_mode)
if not fd then
  if not loop.fs_stat(destination) then
    print("Could not create file " .. destination)
  else
    print("Failed to complete file creation of " .. destination)
  end
else
  loop.fs_chmod(destination, 420)
  loop.fs_close(fd)
end
```

**Comment by fundon** [comment]
It doesn't work.
<img width="422" alt="Screenshot 2023-10-19 at 11 24 28" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/5e12566e-7869-4083-a8be-2c7042fa5a66">

Switched to nvim-tree. It works fine.
<img width="544" alt="Screenshot 2023-10-19 at 11 25 40" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/fc38fa63-1e3d-4cf0-a0f6-7ca9fc5e455a">

**Comment by cseickel** [comment]
Your screenshots show that you are working in completely different directories. The one that worked is in:
```
/Users/fundon
```

The one that does not is in:
```
/home/fundon
```

What happens when you use neo-tree to create a file in `/Users/fundon/Documents/`?

**Comment by fundon** [comment]
Before I make symbolic link:
> 6\. `ln -sf /Users/fundon/Documents $HOME`

**Comment by fundon** [comment]
I tried. 

Tests in `/Users/fundon` and `/Users/fundon/Documents` folders

<img width="742" alt="Screenshot 2023-10-22 at 01 05 15" src="https://github.com/nvim-neo-tree/neo-tree.nvim/assets/27926/dc2ffa21-41eb-4171-b67a-bc3e35b56b43">

**Comment by pysan3** [comment]
@fundon 

What happens if you change `open_mode` to 420?
That's basically what nvim-tree guys are passing to `vim.loop.fs_open` (but I'm not sure why they came up with this value).

```lua
local loop = vim.loop
local destination = vim.fn.fnamemodify("~/Documents/b.txt", ":p")
-- local open_mode = loop.constants.O_CREAT
--   + loop.constants.O_WRONLY
--   + loop.constants.O_TRUNC
local open_mode = 420
local fd = loop.fs_open(destination, "w", open_mode)
if not fd then
  if not loop.fs_stat(destination) then
    print("Could not create file " .. destination)
  else
    print("Failed to complete file creation of " .. destination)
  end
else
  loop.fs_chmod(destination, 420)
  loop.fs_close(fd)
end
```

**Comment by fundon** [comment]
@pysan3 

Thanks, it works.

I check the file permissions:

```
-rw-r--r--  1 fundon users    0 Oct 23 03:56 a.txt
-rw-r--r--  1 fundon users    0 Oct 23 03:55 b.txt
```

`-rw-r--r--` is `0b110100100`, so the mode is `420`.

**Comment by pysan3** [comment]
I thought the 3rd argument to `fs_open` should be values calculated by doing bit-or of `constant.O_WRONLY` etc but does that mean we should pass the permission value?

I now wonder why nobody has noticed this before. 

And if this is true, we don't need to `fs_chmod` afterwards as well. 

Could you double check and fix it @cseickel ?

**Comment by pysan3** [comment]
Thanks for investing @cseickel 

This might fix #911 as well which I couldn't find a solution and fixed by implementing a workaround.
